doc: pgvector compatibility (#14)
tensorchord/pgvecto.rs-docs: src/usage/compatibility.md

# Pgvector compatibility

Pgvecto.rs is natively compatible with pgvector at:
* Create table like `CREATE TABLE t (val vector(3))`
* Query vectors like `INSERT INTO t (val) VALUES ('[0.6,0.6,0.6]')`

> Pgvecto.rs can be configured to be compatible with pgvector at: 

Reviewer: ```suggestion
`pgvecto.rs` can be configured to be compatible with `pgvector` in: 
```

* Create vector indexes like `CREATE INDEX ON t USING hnsw (val vector_ip_ops);`

> * Query options like `SET ivfflat.probes=10;`

Reviewer: ```suggestion
* Query options, which could be set by `SET ivfflat.probes = 10;`
```

This feature is called `compatibility mode`. 

> It can be enabled with `SET vectors.pgvector_compatibility=on`.

Reviewer: ```suggestion
It can be enabled by `SET vectors.pgvector_compatibility = on`.
```

Variables for `ivfflat`:

| Option           | Type                     | Default | Description                               |
| ---------------- | ------------------------ | ------- | ----------------------------------------- |
| ivfflat.probes   | integer (`[1, 1000000]`) | `10`    | Number of lists to scan.                  |

::: warning
Default value of `ivfflat.probes` is `10` instead of `1` from pgvector.
:::

Options for `hnsw`:

| key             | type    | default | description                      |
| --------------- | ------- | ------- | -------------------------------- |
| m               | integer | `16`    | Maximum degree of the node.      |
| ef_construction | integer | `64`    | Search extent in construction.   |

Variables for `hnsw`:

| Option         | Type                     | Default | Description                               |
| -------------- | ------------------------ | ------- | ----------------------------------------- |
| hnsw.ef_search | integer (`[1, 65535]`)   | `100`   | Search scope of HNSW.                     |

::: warning
Default value for `hnsw.ef_search` is `100` instead of `40` from pgvector.
:::

> ::: tip

Reviewer: Remove.

-- [hnsw + vector_cosine_ops] index with single ef_construction option
CREATE INDEX hnsw_cosine_index ON t USING hnsw (val vector_cosine_ops) WITH (ef_construction = 80);
-- anonymous [hnsw + vector_ip_ops] with all options
CREATE INDEX ON t USING hnsw (val vector_ip_ops) WITH (ef_construction = 80, m = 12);
-- [ivfflat + vector_l2_ops] index with default options
CREATE INDEX ivfflat_l2_index ON t USING ivfflat (val vector_l2_ops);
-- [ivfflat + vector_ip_ops] index with all options
CREATE INDEX ivfflat_ip_index ON t USING ivfflat (val vector_cosine_ops) WITH (nlist = 80);
-- anonymous [ivf + vector_ip_ops] with all options
CREATE INDEX ON t USING ivfflat (val vector_ip_ops) WITH (lists = 80)
```

## Limitation

For compatibility, we strive to maintain a consistent user experience, but there are still some limitations in two aspects:

* Some features of pgvector.rs are not available in `compatibility mode'. 
* Some features of pgvector are different in `compatibility mode 

### Inaccessible features of pgvecto.rs
When executing SQL statements in `compatibility mode` without `vectors` index, some features of pgvecto.rs are not accessible:

> * `flat` index 

Reviewer: I think we do not need to list all our features here.

* Quantization, including `scalar quantization` and `product quantization`

> * Prefilter and vbase

Author: My fault, removed.

### Difference from pgvector

> The `compatibility mode` focuses on the most commonly used features for creating indexes. So far, there are still a few differences from pgvector.

Reviewer: ```suggestion
```

> Known problems are not limited to:

Reviewer: ```suggestion
```

> * Create `btree` index on vector type data is not supported

Reviewer: It can be supported by users who want it. So I think we can remove it.

> * Create `btree` index on vector type data is not supported

Reviewer: ```suggestion
* Creating `btree` indexes on a column of `vector` data type is not supported
```

# Compatible with pgvector

pgvecto.rs can be configured to be compatible with `pgvector` at: 
- Create table
- Create vector indexes
- Search vectors

> For `create table` and `search vectors`, pgvecto.rs supports this feature natively.

Reviewer: > pgvecto.rs can be configured to be compatible with `pgvector` at

Since it's native, shouldn't it unconfigurable?

> For `create vector indexes' this feature should be enabled with `SET vectors.pgvector_compat=on;`.

Reviewer: Typo.

```sql
SET vectors.pgvector_compat=on;
-- [hnsw + vector_l2_ops] index with default options
CREATE INDEX hnsw_l2_index ON t USING hnsw (val vector_l2_ops);
-- [hnsw + vector_cosine_ops] index with single ef_construction option
CREATE INDEX hnsw_cosine_index ON t USING hnsw (val vector_cosine_ops) WITH (ef_construction = 80);
-- anonymous [hnsw + vector_ip_ops] with all options
CREATE INDEX ON t USING hnsw (val vector_ip_ops) WITH (ef_construction = 80, m = 12);
-- [ivfflat + vector_l2_ops] index with default options
CREATE INDEX ivfflat_l2_index ON t USING ivfflat (val vector_l2_ops);
-- [ivfflat + vector_ip_ops] index with all options
CREATE INDEX ivfflat_ip_index ON t USING ivfflat (val vector_cosine_ops) WITH (nlist = 80);
-- anonymous [ivf + vector_ip_ops] with all options
CREATE INDEX ON t USING ivfflat (val vector_ip_ops) WITH (lists = 80)
```

## Limitation

For compatibility, we strive to maintain a consistent user experience, but there are still some limitations in two aspects:

- Some features of pgvector.rs are not available in `compatibility mode'. 
- Some features of pgvector are different in `compatibility mode 

### Unavailable features of pgvecto.rs

> When pgvecto.rs is running in `compatibility mode`, some features of pgvecto.rs are not available:

Reviewer: We can use better words to describe the functionality. "compatibility mode" does not harm to features of pgvecto.rs.

Author: Fixed: When executing SQL statements in `compatibility mode', some features of pgvecto.rs are not accessible

- `flat` index
- Quantization, including `scalar quantization` and `product quantization`
- prefilter and vbase

And for index `ivfflat` and `hnsw` only the following options are available. 
Their default value is **different from pgvecto.rs original**, which keeps the same from `pgvector`.

Options for `ivfflat

| Key | Type | Default | Description |
| ---------------- | ------- | ------- | ----------------------------------------- |
| lists | integer | `100` | Number of cluster units.                  |

Options for `hnsw`.

| key | type | default | description |
| --------------- | ------- | ------- | -------------------------------------- |
| m | integer | `16` | Maximum degree of the node.            |
| ef_construction | integer | `64` | Search extent in construction.              |

### Difference from pgvector

Although in `compatibility mode', there are still differences from pgvector.

Known problems are not limited to:
- Create vector indexes is asynchronous at pgvecto.rs, instead of synchronous at pgvector

> - Query Options [ivfflat.probes](https://github.com/pgvector/pgvector#query-options-1) need to be `vectors.ivf_nprobe`, and [hnsw.ef_search](https://github.com/pgvector/pgvector?tab=readme-ov-file#query-options) need to be `vectors.hnsw_ef_search`

Reviewer: Let's put it foremost.

Reviewer 2: I think we decided to add `ivfflat.probes` and `hnsw.ef_search` to pgvecto.rs also

Reviewer: > I think we decided to add `ivfflat.probes` and `hnsw.ef_search` to pgvecto.rs also?

Which GUC will we use for searching?

Reviewer 2: Can we support both? We can keep our gucs in documentations

Reviewer: > Can we support both? We can keep our gucs in documentations

We have to decide the value of which GUC to pass to the bgworker.

Reviewer 2: Can we define two gucs on the same variable?

Reviewer 2: Then we can add an extra if to check whether the compatibility mode is enabled...

Reviewer 2: Basically when we said compatible, we should make it compatible as much as we can

Reviewer: We can just hook `VariableSetStmt` and `VariableShowStmt`.

Reviewer 2: hook is kind of hacky, better not using hook for gucs.

> - [Vector Functions](https://github.com/pgvector/pgvector?tab=readme-ov-file#vector-functions) and [Aggregate Functions](https://github.com/pgvector/pgvector?tab=readme-ov-file#aggregate-functions) are not supported

Reviewer: It's not needed to document.

> # Compatible with pgvector

Reviewer: Is a noun better?

> pgvecto.rs is natively compatibile with pgvector at:

Reviewer: Typo.

# Pgvector compatibility

Pgvecto.rs is natively compatible with pgvector at:

> - Create table

Reviewer: Let's remove it.

Author: But if user want to use this mode, there is nowhere to tell him that `Create table` and `Search vectors` is the same other than here.

- Search vectors

Pgvecto.rs can be configured to be compatible with pgvector at: 
- Create vector indexes

This feature is called `compatibility mode`. 

> It can be enabled with `SET vectors.pgvector_compatibility=on;`.

Reviewer: I prefer enable_* in fact because it keeps same style with PostgreSQL GUCs.

Author: It would be too longer like `enable_pgvector_compatibility`, actually I referred to https://github.com/lanterndata/lantern#a-note-on-operators-and-operator-classes

> ## Examples

Reviewer: We need to talk about what the option does and does not before examples.

It's easy to enable compatibility mode and start a vector search.
```sql
DROP TABLE IF EXISTS t;

> SET vectors.pgvector_compat=on;

Reviewer: Outdated.

CREATE TABLE t (val vector(3));
INSERT INTO t (val) SELECT ARRAY[random(), random(), random()]::real[] FROM generate_series(1, 1000);
CREATE INDEX hnsw_l2_index ON t USING hnsw (val vector_l2_ops);
SELECT COUNT(1) FROM (SELECT 1 FROM t ORDER BY val <-> '[0.5,0.5,0.5]' limit 100) t2;
DROP INDEX hnsw_l2_index;
```

Multiply types of indexes are accepted:
```sql

> SET vectors.pgvector_compat=on;

Reviewer: Outdated.

- Some features of pgvector are different in `compatibility mode 

### Inaccessible features of pgvecto.rs
When executing SQL statements in `compatibility mode', some features of pgvecto.rs are not accessible:
- `flat` index
- Quantization, including `scalar quantization` and `product quantization`
- prefilter and vbase

And for index `ivfflat` and `hnsw` only the following options are available. 
Their default value is **different from pgvecto.rs original**, which keeps the same from `pgvector`.

Options for `ivfflat

| Key | Type | Default | Description |
| ---------------- | ------- | ------- | ----------------------------------------- |
| lists | integer | `100` | Number of cluster units.                  |

Options for `hnsw`.

| key | type | default | description |
| --------------- | ------- | ------- | -------------------------------------- |
| m | integer | `16` | Maximum degree of the node.            |
| ef_construction | integer | `64` | Search extent in construction.              |

### Difference from pgvector

Although in `compatibility mode', there are still differences from pgvector.

Known problems are not limited to:

> - `bree` index

Reviewer 2: ```suggestion
- `btree` index
```

> - `btree` index

Reviewer: Why is it a problem? What's relationship about btree index and pgvector compatibility?

Author: extended to [`btree` index is not supported`].

CREATE INDEX ON t USING ivfflat (val vector_ip_ops) WITH (lists = 80)
```

## Limitation

For compatibility, we strive to maintain a consistent user experience, but there are still some limitations in two aspects:

- Some features of pgvector.rs are not available in `compatibility mode'. 
- Some features of pgvector are different in `compatibility mode 

### Inaccessible features of pgvecto.rs
When executing SQL statements in `compatibility mode` without `vectors` index, some features of pgvecto.rs are not accessible:
- `flat` index 
- Quantization, including `scalar quantization` and `product quantization`
- Prefilter and vbase

### Difference from pgvector

Although in `compatibility mode', there are still differences from pgvector.

Known problems are not limited to:
- Create `btree` index on vector type data is not supported
- Query Options [ivfflat.probes](https://github.com/pgvector/pgvector#query-options-1) need to be `vectors.ivf_nprobe`, and [hnsw.ef_search](https://github.com/pgvector/pgvector?tab=readme-ov-file#query-options) need to be `vectors.hnsw_ef_search`
- Create vector indexes is asynchronous at pgvecto.rs, instead of synchronous at pgvector

> - All other modules except index are not supported

Reviewer: Let's just say "pgvector compatibility" provide compatibility of creating an index at the foremost.
